vecindex: lazily create root partition metadata

andy-kimball opened the pull request:
When preparing to update a K-means tree, we check its root partition metadata. If that does not yet exist, we will create it using a CPut operation and then acquire a SHARED lock that prevents a background split operation until the transaction ends.

Epic: CRDB-42943

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142904)
<!-- Reviewable:end -->

andy-kimball committed:
vecindex: lazily create root partition metadata

When preparing to update a K-means tree, we check its root partition metadata.
If that does not yet exist, we will create it using a CPut operation and then
acquire a SHARED lock that prevents a background split operation until the
transaction ends.

Epic: CRDB-42943

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142904#-:-OLLcBiO0CdcN4eAy1dN:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 327 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142904#-OLLbbmnArrSai31y5nv:-OLLbbmnArrSai31y5nw:b-j6862n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9bc9e7eb8d5d4d3f1ddf24c8cc9a50d37f2cf4b/pkg/sql/vecindex/vecstore/store_txn.go#L327)):*
> ```Go
> 	// If we're preparing to update the root partition, then lazily create its
> 	// metadata if it does not yet exist, and lock that record.
> 	if forUpdate && partitionKey == cspann.RootKey && b.Results[0].Rows[0].Value == nil {
> ```

nit: add a TODO to instead use the `LockNonExisting` on the initial Get once https://github.com/cockroachdb/cockroach/pull/141781 is merged.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 596 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142904#-OLLbpfAAZ3CGImy1LGJ:-OLLbpfAAZ3CGImy1LGK:b-j2jfrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9bc9e7eb8d5d4d3f1ddf24c8cc9a50d37f2cf4b/pkg/sql/vecindex/vecstore/store_txn.go#L596)):*
> ```Go
> 	}
> 	b.CPut(metadataKey, meta, []byte{})
> 	b.GetForShare(metadataKey, tx.lockDurability)
> ```

nit: is the locking `Get` necessary here? The intent placed for the `CPut` should already function as a lock.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142904)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 327 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142904#-OLLbbmnArrSai31y5nv:-OLLl8rBDyJ38N_gckFq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9bc9e7eb8d5d4d3f1ddf24c8cc9a50d37f2cf4b/pkg/sql/vecindex/vecstore/store_txn.go#L327)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: add a TODO to instead use the `LockNonExisting` on the initial Get once https://github.com/cockroachdb/cockroach/pull/141781 is merged.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 596 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142904#-OLLbpfAAZ3CGImy1LGJ:-OLLl9T-DlD5Ub5kq-kf:btljozp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9bc9e7eb8d5d4d3f1ddf24c8cc9a50d37f2cf4b/pkg/sql/vecindex/vecstore/store_txn.go#L596)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: is the locking `Get` necessary here? The intent placed for the `CPut` should already function as a lock.
</blockquote></details>

You're right. The GetForShare is redundant in this case. If another writer has won the race, then `Run` will fail with a `WriteTooOld` error; there's no way that `GetForShare` can acquire a shared lock on the metadata row, since its timestamp will be higher.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
*[`pkg/sql/vecindex/vecstore/store_txn.go` line 596 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142904#-OLLbpfAAZ3CGImy1LGJ:-OLMGKPB-Yhanzx8n3Aw:b-uyx2g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9bc9e7eb8d5d4d3f1ddf24c8cc9a50d37f2cf4b/pkg/sql/vecindex/vecstore/store_txn.go#L596)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You're right. The GetForShare is redundant in this case. If another writer has won the race, then `Run` will fail with a `WriteTooOld` error; there's no way that `GetForShare` can acquire a shared lock on the metadata row, since its timestamp will be higher.
</blockquote></details>

I added a test to ensure that racing writers will either not error or fail with `WriteTooOld`.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r=drewkimball
